packages/cli/src/execution-lifecycle-hooks/execution-hooks-factory.ts

```diff
@@ -0,0 +1,464 @@
+		// So to avoid confusion, we are removing other hooks.
+
+		const { executionRepository } = this;
+		// TODO: >>> clear all nodeExecuteAfter hooks <<<
+		// hookFunctions.nodeExecuteAfter = [];
+		hooks.addHook('workflowExecuteAfter', async function (fullRunData) {
+			// Don't delete executions before they are finished
+			if (!fullRunData.finished) return;
+
+			const executionStatus = determineFinalExecutionStatus(fullRunData);
+			fullRunData.status = executionStatus;
+
+			const saveSettings = toSaveSettings(this.workflowData.settings);
+
+			const shouldNotSave =
+				(executionStatus === 'success' && !saveSettings.success) ||
+				(executionStatus !== 'success' && !saveSettings.error);
+
+			if (shouldNotSave) {
+				await executionRepository.hardDelete({
+					workflowId: this.workflowData.id,
+					executionId: this.executionId,
+				});
+			}
+		});
+
+		return hooks;
+	}
+
+	/** Returns ExecutionHooks instance for running sub-workflows */
+	forSubExecution(
+		mode: WorkflowExecuteMode,
+		executionId: string,
+		workflowData: IWorkflowBase,
+		optionalParameters: ExecutionHooksOptionalParameters,
+	) {
+		const hooks = new ExecutionHooks(mode, executionId, workflowData, optionalParameters);
+		this.addPreExecuteHooks(hooks);
+		this.addEventHooks(hooks);
+		this.addSavingHooks(hooks, false);
+		return hooks;
+	}
+
+	private addPreExecuteHooks(hooks: ExecutionHooks) {
+		const { externalHooks } = this;
+
+		hooks.addHook('workflowExecuteBefore', async function (workflow) {
+			await externalHooks.run('workflow.preExecute', [workflow, this.mode]);
+		});
+
+		// TODO: skip this if saveSettings.progress is not true
+		hooks.addHook('nodeExecuteAfter', async function (nodeName, data, executionData) {
+			await saveExecutionProgress(
+				this.workflowData,
+				this.executionId,
+				nodeName,
+				data,
+				executionData,
+				this.pushRef,
+			);
+		});
+	}
+
+	private addEventHooks(hooks: ExecutionHooks) {
+		const { eventService, workflowStatisticsService } = this;
+		hooks.addHook('nodeExecuteBefore', async function (nodeName) {
+			const { executionId, workflowData: workflow } = this;
+			eventService.emit('node-pre-execute', { executionId, workflow, nodeName });
+		});
+
+		hooks.addHook('nodeExecuteAfter', async function (nodeName) {
+			const { executionId, workflowData: workflow } = this;
+			eventService.emit('node-post-execute', { executionId, workflow, nodeName });
+		});
+
+		hooks.addHook('workflowExecuteBefore', async function () {
+			const { executionId, workflowData } = this;
+			eventService.emit('workflow-pre-execute', { executionId, data: workflowData });
+		});
+
+		hooks.addHook('nodeFetchedData', async (workflowId, node) => {
+			workflowStatisticsService.emit('nodeFetchedData', { workflowId, node });
+		});
+	}
+
+	/** Returns hook functions to save workflow execution and call error workflow */
+	private addSavingHooks(hooks: ExecutionHooks, isWorker: boolean) {
```

Comment:
We'll need much more work until hooks are organized in a way that makes sense, but until we get there maybe we can use more accurate names, even rather long.

```suggestion
	private addWorkflowExecuteAfterCallback(hooks: ExecutionHooks, isWorker: boolean) {
```